bigtable/integration_test.go

```diff
@@ -2142,6 +2144,63 @@ func TestIntegration_AdminBackup(t *testing.T) {
 	}
 }
 
+// Blackhole directpath address to test fallback.
+func TestIntegration_DirectPathFallback(t *testing.T) {
+	ctx := context.Background()
+	testEnv, _, _, table, _, cleanup, err := setupIntegration(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
+
+	if err := populatePresidentsGraph(table); err != nil {
+		t.Fatal(err)
+	}
+
+	// Precondition: wait for DirectPath to connect.
+	countEnough := exerciseDirectPath(ctx, testEnv, table /*blackholeDP = */, false)
+	if !countEnough {
+		t.Fatalf("Failed to observe RPCs over DirectPath")
+	}
+
+	// Enable the blackhole, which will prevent communication with grpclb and thus DirectPath.
+	blackholeOrAllowDirectPath(testEnv, t /*blackholeDP = */, true)
+	countEnough = exerciseDirectPath(ctx, testEnv, table /*blackholeDP = */, true)
+	if !countEnough {
+		t.Fatalf("Failed to fallback to CFE after blackhole DirectPath")
+	}
+
+	// Make sure that the client will start reading from IPv6 again by sending new requests and
+	// checking the injected IPv6 counter has been updated.
+	blackholeOrAllowDirectPath(testEnv, t /*blackholeDP = */, false)
+	countEnough = exerciseDirectPath(ctx, testEnv, table /*blackholeDP = */, false)
+	if !countEnough {
+		t.Fatalf("Failed to fallback to CFE after blackhole DirectPath")
+	}
+}
+
+func exerciseDirectPath(ctx context.Context, testEnv IntegrationEnv, table *Table, isBlackhole bool) bool {
+	var numCount uint64
+	const (
+		numRPCsToSend  = 20
+		minCompleteRPC = 40
+	)
+
+	countEnough := false
+	start := time.Now()
+	for !countEnough && time.Since(start) < 2*time.Minute {
+		for i := 0; i < numRPCsToSend; i++ {
+			_, _ = table.ReadRow(ctx, "jadams")
+			if _, useDp := isDirectPathRemoteAddress(testEnv); useDp != isBlackhole {
+				atomic.AddUint64(&numCount, 1)
```

Comment:
I think I'm missing something -- is this run in parallel?

Reply:
The bigtable client will open 4 channels by default (https://github.com/googleapis/google-cloud-go/blob/master/bigtable/bigtable.go#L78), so they could be run in parallel.

Reply:
This particular code will not run in parallel. The code you linked to is for pooling the underlying network connections.

Reply:
Thanks for pointing out. I removed the atomic operation.